uk.ac.kcl.inf.modelling.xdsml.gemoc_henshin/src/uk/ac/kcl/inf/modelling/xdsml/gemoc_henshin/engine/ui/launcher/HenshinConcurrentEngineLauncher.xtend

```diff
@@ -0,0 +1,290 @@
+
+			// detect if we are running in debug mode or not
+			var ExecutionMode executionMode = null;
+			if (ILaunchManager.DEBUG_MODE.equals(mode)) {
+				executionMode = ExecutionMode.Animation;
+			} else {
+				executionMode = ExecutionMode.Run;
+			}
+
+			// stop the launch if an engine is already running for this model
+			if (isEngineAlreadyRunning(runConfiguration.getExecutedModelURI())) {
+				return;
+			}
+			
+			//!!!!!!!!
+			//flag to handle concurrent steps 
+			//!!!!!!!
+			var showSequenceRules = true;
+			var HenshinConcurrentModelExecutionContext concurrentexecutionContext = new HenshinConcurrentModelExecutionContext(
+					runConfiguration, executionMode, showSequenceRules);
+					 
+			concurrentexecutionContext.initializeResourceModel();
+					
+			//initialize the solver
+			var HenshinSolver _solver
+			try {
+				_solver = concurrentexecutionContext.getLanguageDefinitionExtension().instanciateSolver() as HenshinSolver;
+				_solver.initialize(concurrentexecutionContext);
+			} catch (CoreException e) {
+				throw new CoreException(new Status(Status.ERROR, Activator.PLUGIN_ID,
+						"Cannot instanciate solver from language definition", e));
+			}
+	
+			_executionEngine = new HenshinConcurrentExecutionEngine(concurrentexecutionContext, _solver);
+			
+			openViewsRecommandedByAddons(runConfiguration);
+
+			//define a new job to run the engine
+			val Job job = new Job(getDebugJobName(configuration, getFirstInstruction(configuration))) {
+				override protected IStatus run(IProgressMonitor monitor) {
+					//debug mode with Sirius animator
+					if (ILaunchManager.DEBUG_MODE.equals(mode)) {
+						val IEngineAddon animator = AbstractGemocAnimatorServices.getAnimator();
+						var plat = _executionEngine.getExecutionContext().getExecutionPlatform()					
+						plat.addEngineAddon(animator);
+						try {
+							HenshinConcurrentEngineLauncher.super.launch(configuration, mode, launch, monitor);
+							return new Status(IStatus.OK, getPluginID(), "Execution was launched successfully");
+						} catch (CoreException e) {
+							e.printStackTrace();
+							return new Status(IStatus.ERROR, getPluginID(), "Could not start debugger.");
+						}
+					}
+					// if not debug mode, start the engine
+					else {
+						_executionEngine.start();
+						debug("Execution finished.");
+						return new Status(IStatus.OK, getPluginID(), "Execution was launched successfully");
+					}
+				}
+			};
+			
+			debug("Initialization done, starting engine...");
+			job.schedule();
+
+		} catch (Exception e) {
+			val String message = "Error occured when starting execution engine: " + e.getMessage()
+					+ " (see inner exception).";
+			// error(message);
+			Activator.error(message, e);
+			throw new CoreException(new Status(Status.ERROR, Activator.PLUGIN_ID, message, e));
+		}
+	}
+	
+	//the rest of methods below are copied implementations from the GEMOC Concurrent Engine
```

Comment:
Why do we have to copy them? Can we possibly inherit them from their code? If not, should we submit an issue for them to refactor and make this available?

Reply:
there seems to be no generic launcher. there's only one linked so the CSSL engine that's why I've copied that code. Perhaps they could provide a more generic version.